rbx_types/src/attributes/mod.rs

```diff
@@ -0,0 +1,259 @@
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Reads from a serialized attributes string, and produces a new `Attributes` from it.
+    pub fn from_reader<R: Read>(reader: R) -> Result<Self, Error> {
+        Ok(Attributes {
+            data: get_attributes(reader)?,
+        })
+    }
+
+    /// Writes the attributes as a serialized string to the writer.
+    pub fn to_writer<W: Write>(&self, mut writer: W) -> Result<(), Error> {
+        let bytes = attributes_from_map(self.data.iter())?;
+        writer
+            .write_all(&bytes)
+            .map_err(AttributeError::ToWriterFail)
+            .map_err(Into::into)
+    }
+
+    /// Get the attribute with the following key.
+    pub fn get<K: Borrow<str>>(&self, key: K) -> Option<&Variant> {
+        self.data.get(key.borrow())
+    }
+
+    /// Inserts an attribute with the given key and value.
+    /// Will return the attribute that used to be there if one existed.
+    pub fn insert(&mut self, key: String, value: Variant) -> Option<Variant> {
+        self.data.insert(key, value)
+    }
+
+    /// Removes an attribute with the given key.
+    /// Will return the value that was there if one existed.
+    pub fn remove<K>(&mut self, key: K) -> Option<Variant>
+    where
+        K: Hash + Eq,
+        String: Borrow<K>,
+    {
+        self.data.remove(&key)
+    }
+
+    /// Returns an iterator of borrowed attributes.
+    pub fn iter(&self) -> impl Iterator<Item = (&String, &Variant)> {
+        self.data.iter()
+    }
+}
+
+impl IntoIterator for Attributes {
+    type IntoIter = AttributesIntoIter;
+    type Item = (String, Variant);
+
+    fn into_iter(self) -> Self::IntoIter {
+        AttributesIntoIter {
+            iter: self.data.into_iter(),
+        }
+    }
+}
+
+impl FromIterator<(String, Variant)> for Attributes {
+    fn from_iter<T: IntoIterator<Item = (String, Variant)>>(iter: T) -> Self {
+        Self {
+            data: iter.into_iter().collect(),
+        }
+    }
+}
+
+/// An owning iterator over the entries of an `Attributes`.
+/// This is created by [`Attributes::into_iter`].
+pub struct AttributesIntoIter {
+    iter: std::collections::hash_map::IntoIter<String, Variant>,
+}
+
+impl Iterator for AttributesIntoIter {
+    type Item = (String, Variant);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next()
+    }
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum AttributeError {
+    #[error("invalid value type: {0}")]
+    InvalidValueType(u8),
+
+    #[error("invalid brick color: {0}")]
+    InvalidBrickColor(u32),
+
+    #[error("invalid name")]
+    InvalidName,
+
+    #[error("invalid size")]
+    InvalidSize,
+
+    #[error("invalid variant type passed")]
+    InvalidVariantType,
+
+    #[error("no value type was found")]
+    NoValueType,
+
+    #[error("malformed attribute key")]
+    MalformedEntryKey(FromUtf8Error),
+
+    #[error("couldn't write to writer: {0}")]
+    ToWriterFail(io::Error),
```

Comment:
When tracking a source error in Rust, it's unidiomatic to include the source error in your error's message. Instead, thiserror has a `#[source]` attribute (which is automatically applied to any fields named `source`) that does the right thing for you!

Reply:
What *is* the right thing out of curiosity? What does `#[source]` do?

Reply:
The correct thing to do is to write an error message without including the inner error, like "couldn't write to writer" or something more specific.

Adding `#[source]` makes the generated `std::error::Error` implementation have a correct value for [`Error::source`](https://doc.rust-lang.org/stable/std/error/trait.Error.html#method.source). When the error is printed with debug, the chain of source errors will be printed out!